### PR TITLE
Baremetal: Add diagnostic error message for ironic terraform errors

### PR DESCRIPTION
**Pull request by kirankt**
baremetal: Add diagnostic error message for ironic terraform errors

This PR enables the installer to track certain ironic terraform provider error messages and translates it into easy-to-understand errors output by the installer.

Examples:
Terraform errors such as "Error: could not contact Ironic API: timeout reached" and "Error: could not inspect: could not inspect node, node is currently 'inspect failed', last error was 'timeout reached while inspecting the node" will each get translated to their corresponding user-friendly error messages.

### CODE REVIEW

File: pkg/terraform/diagnose.go [hunk 1]
```diff
@@ -78,4 +78,9 @@ var conditions = []condition{{
 
 	reason:  "GCPComputeBackendTimeout",
 	message: `GCP is experiencing backend service interuptions, the compute instance failed to create in reasonable time.`,
+}, {
+	match: regexp.MustCompile(`Error: could not contact API: .*`),
+
+	reason:  "BaremetalIronicAPITimeout",
+	message: `Timed out connecting to Ironic API. Please make sure that Ironic API service is up and running.`,
```
Comment (stbenjam) [inline pkg/terraform/diagnose.go:85]: I think we should not assume a user knows what Ironic is (we try to hide it as much as possible), and give them some more details:

*Timed out waiting for provisioning service. This failure can be caused by misconfiguration or inability to download the machine operating system images. Please check the bootstrap host for failing services.*

Coupled with https://github.com/openshift/installer/pull/3156, users will get a list of services to check when connecting to the bootstrap.

Comment (kirankt) [reply pkg/terraform/diagnose.go:85]: Awesome. This makes a lot of sense. Will update.

File: pkg/terraform/diagnose.go [hunk 2]
```diff
@@ -78,4 +78,9 @@ var conditions = []condition{{
 
 	reason:  "GCPComputeBackendTimeout",
 	message: `GCP is experiencing backend service interuptions, the compute instance failed to create in reasonable time.`,
+}, {
+	match: regexp.MustCompile(`Error: could not contact API: .*`),
```
Comment (abhinavdahiya) [inline pkg/terraform/diagnose.go:82]: This regex match is a little too vague to not pick up non ironic failies imo.

Comment (kirankt) [reply pkg/terraform/diagnose.go:82]: It currently matches two error messages that look like these:
```
level=error msg="Error: could not contact API: timeout reached"
level=error
level=error msg="  on ../../../../tmp/openshift-install-431515935/masters/main.tf line 1, in resource \"ironic_node_v1\" \"openshift-master-host\":"
level=error msg="   1: resource \"ironic_node_v1\" \"openshift-master-host\" {"
level=error
level=error
level=error
level=error msg="Error: could not contact API: timeout reached"
level=error
level=error msg="  on ../../../../tmp/openshift-install-431515935/masters/main.tf line 1, in resource \"ironic_node_v1\" \"openshift-master-host\":"
level=error msg="   1: resource \"ironic_node_v1\" \"openshift-master-host\" {"
level=error
level=error
level=error
level=error msg="Error: could not contact API: context deadline exceeded"
level=error
level=error msg="  on ../../../../tmp/openshift-install-431515935/masters/main.tf line 1, in resource \"ironic_node_v1\" \"openshift-master-host\":"
level=error msg="   1: resource \"ironic_node_v1\" \"openshift-master-host\" {"
level=error
```

Comment (stbenjam) [reply pkg/terraform/diagnose.go:82]: Yea, this is the maximum match. We can fix this in the terraform provider: https://github.com/openshift-metal3/terraform-provider-ironic/pull/45

Comment (stbenjam) [reply pkg/terraform/diagnose.go:82]: That's been merged and I released v0.2.3. You should be able to bump the go.mod

Comment (kirankt) [reply pkg/terraform/diagnose.go:82]: Thanks. Won't we still have two different errors for the Ironic API errors? One from context and the other one we print explicitly?

Comment (kirankt) [reply pkg/terraform/diagnose.go]: Added examples in the commit and PR messages.

**Comment by stbenjam** [comment]
/lgtm
/hold

This looks good to me. Remove the hold once e2e-metal-ipi passes.

**Pull request merged**